src/metrics.cpp

```diff
@@ -86,6 +86,7 @@ boost::synchronized_value<std::list<uint256>> trackedBlocks;
 
 boost::synchronized_value<std::list<std::string>> messageBox;
 boost::synchronized_value<std::string> initMessage;
+boost::synchronized_value<std::string> initMessageAfterLoading;
```

Comment:
We have two different threads acting (in the sense of setting status and messages, a third thread running `ThreadShowMetricsScreen` is responsible of actually printing to GUI) here :

a->`AppInit2` (main thread)

1a)   "Verifying wallet..."
2a)   "Loading block index..."
3a)   "Verifying blocks..."
4a)   "Loading wallet..."
5a)   "Activating best chain..."
5.5a) ThreadImport starts on another thread
6a)   "Done loading" (via `SetDoneLoadingUI`)
7a)   no printing if not importing, otherwise print importing info

b->`ThreadImport` (another thread)
1b)   "Reindexing block headers from files...(blk%05u.dat)"
2b)   "Reindexing block from files...(blk%05u.dat)"

Now, due to the synchronization mechanism:
```
while (!fRequestShutdown && chainActive.Tip() == NULL)
    MilliSleep(10);
```
the actual sequence of events is (considering for example 3 blkXXXXX.dat available):

...
4a)
5a)
1b)  blk00000.dat, blk00001.dat, blk00002.dat
2b*) blk00000.dat
6a)
2b**) blk00001.dat, blk00002.dat

Hence, if `initMessageAfterLoading` would not be present (having only `initMessage`):
1. the step 2b* would be immediately overwritten, with "Done Loading"
2. after printing "Done Loading" the GUI would not print `initMessage` (resulting in the user feeling all the loading has been completed and no importing is underway, guessing the possible blockchain progress due to p2p download)
3. after some minutes (this is the time required for importing a blkXXXXX.dat file; full importing, not headers only), step 2b** would trigger an update of `initMessage`

Conversely, introducing `initMessageAfterLoading`, the messages are stored in two different strings and the decision related to which string should be printed is based on the loading/importing status.

Reply:
I agree with @a-petrini : all additions are not needed here.
This
```
while (!fRequestShutdown && chainActive.Tip() == NULL)
    MilliSleep(10);
```
actually suspends main thread until all reindexing finishes i.e. after InitBlockIndex is completely finalized (which instead does nothing if either fReindex or fReindexFast are true)